### DISCUSSION TITLE
A question about wrong pieces and right pieces

### DISCUSSION THREAD

**Discussion by arius28** [flbulgarelli/headbreaker#34] [Q&A] [answered]
Hey @flbulgarelli,

Sorry for asking but what would be an easy way to validate the pieces in a basic scenario with an image and autogenerating the pieces. For example I have a DONE button and when clicked I would like to say : You connected 10 pieces right and 5 wrong. 

I see there is a piecevalidator but i cannot grasp how to use it for such a scenario. I thought of going with a foreach through all the puzzle pieces and checking the connections but wanted to see if there is a better way. I tried to browse through the tests source code to figure out how to use refs but couldn't reach a conclusion. 

Thank you in advance

**Comment by Faycal88** [comment]
@arius28 
May i ask if you have found a solution to that ?

**Comment by flbulgarelli** [comment, marked as answer]
Hi! As @arius28 pointed out, you can use the `PieceValidator`, but in your case you will need a custom validator instead of for example the one you can attach using `attachSolvedValidator`. 

I agree with @arius28 in that probably iterating over the pieces is the best way, since you seem to need to check for the count of properly connected instead of just whether they are fully properly connected. Actually the builtin validator does something similar to that: 

https://github.com/flbulgarelli/headbreaker/blob/c5bd65720fdb091163058ff6b873a6091b7f780f/src/spatial-metadata.js#L30

https://github.com/flbulgarelli/headbreaker/blob/c5bd65720fdb091163058ff6b873a6091b7f780f/src/spatial-metadata.js#L35-L38

For instance, you could use `filter + length` instead of `every` in order to count the number of pieces that are in their - relative - place. If you also want to count the number of pieces _connected_ in they proper place and / or the number of properly connected connection, that may be a bit more tricky - since you will have to check their `up`, `left`, etc connections, not just the relative position.
